### PR TITLE
Add primitive PRIM_RESOLVES to test if a given call resolves

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
Add primitive PRIM_RESOLVES to test if a given call resolves (#19029)

Add a new primitive called `PRIM_RESOLVES`. It can be used to test
if a given expression resolves. Only call expressions may be passed
as actuals for now, but more expression types could be supported
in the future.

For example:

```chapel
record r1 {
  proc doSomething(a) { writeln(a); }
}

record r2 {
  var inner = new r1();
}

var x = new r2();
param yes = __primitive("resolves", x.inner.doSomething(a=0));
assert(yes);
```

The expression passed as an actual argument to `PRIM_RESOLVES` is not
normalized, which means that any expression may be passed as an
actual, and the expression may be written exactly as it would appear
elsewhere in Chapel code.

Unlike `PRIM_CALL_RESOLVES` and `PRIM_METHOD_CALL_RESOLVES`, a call
passed to `PRIM_RESOLVES` should be able to contain actuals with any
combination of storage kind (e.g. `var`, `param`, `type`).

This means that `PRIM_RESOLVES` can be used to enable `Reflection`
functions to support `type` arguments (see #11802).

In addition, uses of `PRIM_RESOLVES` in internal modules enable
more internal modules to be parsed and converted by the new
compiler.

Uses of `PRIM_METHOD_CALL_RESOLVES` and `PRIM_CALL_RESOLVES` within
internal modules have been replaced with calls to `PRIM_RESOLVES`.

ISSUES

  - The current implementation cannot resolve nested calls
     when the return type of the callee is inferred

FUTURE WORK

  - Expand `PRIM_RESOLVES` to work with any expression type
  - Discuss whether or not all callees should be resolved
  - Add a new procedure to the `Reflection` module which makes use
    of `PRIM_RESOLVES`

TESTING

  - [x] `ALL` on `linux64` when `COMM=none`
  - [x] `ALL` on `linux64` when `COMM=gasnet`

Reviewed by @mppf. Thanks!

Signed-off-by: David Longnecker <dlongnecke-cray@users.noreply.github.com>

### CODE REVIEW

File: compiler/resolution/functionResolution.cpp [hunk 1]
```diff
@@ -8732,10 +8732,34 @@ static Expr* resolveTypeOrParamExpr(Expr* expr) {
...
+// Non-normalizable expressions may nest at an arbitrary depth (because
+// they are not normalized), so we have to search upwards in the tree
+// to look for them.
+static Expr* handleNonNormalizableExpr(Expr* expr) {
+
+  // TODO: Ways to limit our search depth?
+  if (Expr* nonNormalRoot = partOfNonNormalizableExpr(expr)) {
+    if (CallExpr* call = toCallExpr(nonNormalRoot)) {
+      if (call->isPrimitive(PRIM_RESOLVES)) {
+        return preFold(call);
```
Comment (mppf) [inline compiler/resolution/functionResolution.cpp:8744]: Maybe add a comment saying that `preFold` will completely replace `PRIM_RESOLVES` primitive calls, and as a result other elements (like resolveExpr or postFold) are not needed here

File: compiler/resolution/preFold.cpp [hunk 2]
```diff
@@ -1884,6 +1885,81 @@ static Expr* preFoldPrimOp(CallExpr* call) {
...
+    INT_ASSERT(call && call->numActuals() == 1);
+
+    auto exprToResolve = call->get(1);
+
+    // TODO: To resolve an arbitrary expression, we need a way to back out
+    // of resolving any expression without emitting errors. This is
+    // probably easier to do in a new compiler world.
+    if (!isCallExpr(exprToResolve)) {
+      INT_FATAL("PRIM_RESOLVES can only resolve CallExpr for now");
+    }
```
Comment (mppf) [inline compiler/resolution/preFold.cpp:1901]: Doesn't the code below already handle this? At least it handles nested calls, right?

File: compiler/resolution/preFold.cpp [hunk 3]
```diff
@@ -1884,6 +1885,81 @@ static Expr* preFoldPrimOp(CallExpr* call) {
...
+        FnSymbol* resolvedFn = tryResolveCall(call, doResolveCalledFns);
+
+        if (!resolvedFn && !isPartialCall) {
+          didResolveExpr = false;
+          break;
+        }
+
+      // TODO: Way to keep type checking from issuing errors?
+      } else {
+        expr = resolveExpr(expr);
```
Comment (mppf) [inline compiler/resolution/preFold.cpp:1947]: By far the most common case here is PRIM_MOVE / PRIM_ASSIGN setting the type of the LHS. You could add some logic to those paths, but I'm not sure how much effort on this is needed in the near term.

**Review by mppf — APPROVED** [review]

**Comment by dlongnecke-cray** [comment]
There were 132 test failures on my COMM=none run, mostly related to a `__primitive("resolves")` call in `ArrayViewSlice` complaining about a recursive call:


```chapel
    // ArrayViewSlice.chpl
    iter these(param tag: iterKind) ref
      where tag == iterKind.standalone && !localeModelHasSublocales &&
           __primitive("resolves", privDom.these(tag)) {
```
```
$CHPL_HOME/modules/internal/ArrayViewSlice.chpl:349: error: unable to resolve return type of function 'privDom'
$CHPL_HOME/modules/internal/ArrayViewSlice.chpl:178: In iterator 'these':
$CHPL_HOME/modules/internal/ArrayViewSlice.chpl:180: error: called recursively at this point
```

In the worst case I can revert this call to `__primitive("method call resolves")`. I'll spend a small amount of time debugging first.

**Comment by mppf** [comment]
I suspect it has to do with privDom being a parentheses-less method.  I think parentheses-less methods might be resolved during scope resolve and not regular resolution. Can you change the current `proc privDom` to a `inline proc getPrivDom()` and then add `inline proc privDom { return getPrivDom(); }` and see if that fixes it?

**Comment by dlongnecke-cray** [comment]
I tried that but no luck. I think the issue might actually be that the return type of `privDom` is inferred, and so when the compiler tries to resolve one of the normalized moves it can't because `privDom` has a unknown return type unless we fully resolve the body first.

If I add a return type to `privDom` then everything works:

```chapel
inline proc privDom: dom.type {
```

So this raises the question, how should we handle code like this?

```chapel
inline proc privDom {
// ...
return dom;
}

// Cannot resolve moves for temp storing `privDom` unless I resolve definition of `privDom`
if __primitive("resolves", privDom.these(tag)) then ...;
```

- One option is to just always resolve callees, but I worry about this because the internal modules use `PRIM_METHOD_CALL_RESOLVES` and not `PRIM_METHOD_CALL_AND_FN_RESOLVES` and I don't know why
- I cannot return false because that will cause existing programs to break
- I think that I actually could fully resolve callees for sub-expressions? Because for `PRIM_METHOD_CALL_RESOLVES` `privDom` is already a resolved temp, which means the body of `privDom` was resolved

**Comment by mppf** [comment]
For now,  I think you should go with adding the return type to `privDom`. In the future, I would expect that `PRIM_RESOLVES` would resolve the return type (which often means resolving the body of the function) when that return type is needed to be able to resolve the type of something inside of the `PRIM_RESOLVES`.

> One option is to just always resolve callees, but I worry about this because the internal modules use PRIM_METHOD_CALL_RESOLVES and not PRIM_METHOD_CALL_AND_FN_RESOLVES and I don't know why

The thing to think about here is - what if there is a type error in the function you are calling? Historically we have been of the view that it's not canResolve's business to know there is a type error (because before using canResolve for some of these things we used a different system where type errors in the function body caused it to return `false` and then never emit the error, which led to silently failing certain things, performance problems, etc). So if there is a type error inside of the called function, I think we want an error emitted, rather than just returning `false`. If we return `true` and emit a type error that is OK with me though - as long as we emit the error somewhere.

**Comment by dlongnecke-cray** [comment]
When using `PRIM_METHOD_CALL_RESOLVES` (caller, but not callee) on the  expression `privDom.these(tag)`, the call & callee for `privDom` is already resolved when the primitive is folded (because the thing in the primitive call is actually the call temp). So if we just only fully resolved callees in subexpressions (and just make sure not to resolve the callee for `.these(tag)` then it would mimic the behavior that `PRIM_METHOD_CALL_RESOLVES` has today.

Unless that we think that that behavior is wrong and `PRIM_METHOD_CALL_RESOLVES` shouldn't be resolving callees in sub-expressions either?

Also should I put a TODO or note about not being able to resolve everything somewhere?

**Comment by mppf** [comment]
What you are describing for PRIM_METHOD_CALL_RESOLVES probably doesn't make sense for PRIM_RESOLVES because it is intended to ask about if a whole expression resolves, right?

You can leave the TODO in the PR message.

File: compiler/resolution/preFold.cpp (hunk 2 above)
Comment (dlongnecke-cray) [reply compiler/resolution/preFold.cpp:1901]: I added a `(e.g. a SymExpr)` and noted that we'd have to interact with `scopeResolve` as well.

Comment (mppf) [reply compiler/resolution/preFold.cpp:1901]: Do you mean `UnresolvedSymExpr` ? Because a SymExpr is already resolved, by nature.

**Comment by dlongnecke-cray** [comment]
I'll do that, and I also added a future to capture the issue.

**Comment by dlongnecke-cray** [comment]
TODO:

- More test failures, most can be fixed by annotating internal module procedures with return types, I think
- One test failure to do with `--verify`

**Comment by dlongnecke-cray** [comment]
Just got the `COMM=none` tests to pass. 🎉

**Pull request merged**